File: Coding_Problems_Java/797 All Paths From Source to Target.java
class Solution {
    public List<List<Integer>> allPathsSourceTarget(int[][] graph) {
        int n = graph.length;
        List<List<Integer>> res = new ArrayList<>();
        ArrayList<Integer> path = new ArrayList<>();

        dfs(0, graph, path, res, n);
        return res;
    }
    private void dfs(int node, int[][] graph, List<Integer> path, List<List<Integer>> res, int n){
    
        if(node == n-1){
            path.add(node);
            res.add(new ArrayList<>(path));
            path.remove(path.size()-1);
            return;
        }
        path.add(node);
        for(int adjNode: graph[node]){
            dfs(adjNode, graph, path, res, n);
        }
        path.remove(path.size()-1);
    }
}

File: Coding Problems - Python/Graphs/797 All Paths From Source to Target.py
class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        def dfs(node, path):
            if node == n-1:
                res.append(path + [node])
                return

            for adj_node in graph[node]:
                dfs(adj_node, path + [node])
        n = len(graph)
        res = []
        path = []

        dfs(0, path)
        return res
